redis/docs/token_authentication.md

```diff
@@ -0,0 +1,225 @@
+    "your-tenant-id".to_string(),
+    "your-client-id".to_string(),
+    "your-client-secret".to_string(),
+)?;
+```
+
+### Service Principal with Certificate
+
+For enhanced security:
+
+```rust
+use redis::ClientCertificate;
+use std::fs;
+
+// Load certificate from file
+let certificate_base64 = fs::read_to_string("path/to/base64_pkcs12_certificate")
+    .expect("Base64 PKCS12 certificate not found.")
+    .trim()
+    .to_string();
+
+// Create the credentials provider using service principal with client certificate
+let provider = EntraIdCredentialsProvider::new_client_certificate(
+    "your-tenant-id".to_string(),
+    "your-client-id".to_string(),
+    ClientCertificate {
+        base64_pkcs12: certificate_base64, // Base64 encoded PKCS12 data
+        password: None,
+    },
+)?;
+```
+
+### Managed Identity
+
+For Azure-hosted applications:
+
+```rust
+use redis::{ManagedIdentityCredentialOptions, UserAssignedId};
+
+// System-assigned managed identity
+let provider = EntraIdCredentialsProvider::new_system_assigned_managed_identity()?;
+
+// User-assigned managed identity
+let provider = EntraIdCredentialsProvider::new_user_assigned_managed_identity()?;
+// or with custom scopes and identity specification
+let provider = EntraIdCredentialsProvider::new_user_assigned_managed_identity_with_scopes(
+    vec!["your-scope".to_string()],
+    Some(ManagedIdentityCredentialOptions {
+        // Specify the user-assigned identity using one of:
+        user_assigned_id: Some(UserAssignedId::ClientId("your-client-id".to_string())),
+        // or: user_assigned_id: Some(UserAssignedId::ObjectId("your-object-id".to_string())),
+        // or: user_assigned_id: Some(UserAssignedId::ResourceId("your-resource-id".to_string())),
+        ..Default::default()
+    }),
+)?;
+```
+
+## Advanced Configuration
+
+### TokenRefreshConfig
+The `TokenRefreshConfig` allows further customization of the token refresh behavior, based on the token's expiration time, when applicable.
+
+- `expiration_refresh_ratio`: Fraction of token lifetime before refresh (0.0-1.0)
+- `retry_config`: Configuration for retry behavior on failures
+
+### RetryConfig
+
+Configuration for handling failed token refresh attempts.
+
+- `max_attempts`: Maximum number of retry attempts for token refresh.
+- `initial_delay`: Initial delay before attempting to refresh the token after a failure. Subsequent retries use exponential backoff based on this value.
+- `max_delay`: Upper bound for retry delays to prevent excessively long waits. Delays will never exceed this value, even with exponential backoff.
+- `backoff_multiplier`: Growth factor for exponential backoff (typically 2.0 for doubling). Each retry delay is multiplied by this value: `delay = delay * backoff_multiplier`.
+- `jitter_percentage`: Random variation added to delays as a fraction of the calculated delay (0.0 to 1.0). For example, 0.5 means up to ±50% variation to prevent synchronized retries.
```

Comment:
once you integrate this into the rustdoc, there won't be a need to re-document each type, just link to the type's documentation.